### PR TITLE
update DeleteVMs() logic, add support for NetworkSecurityGroupRules, refine logic for testing with TiP

### PR DESCRIPTION
**Pull request by harz566**
04/27/2020 19:16:49 : [INFO ] Test MT98 finished
04/27/2020 19:16:49 : [INFO ] 
[LISAv2 Test Results Summary]
Test Run On           : 04/27/2020 19:01:52
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : Latest
Initial Kernel Version: 5.0.0-1036-azure
Final Kernel Version  : 5.0.0-1036-azure
Total Test Cases      : 3 (3 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:14

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 NETWORK              NET-IFUP-IFDOWN                                                                   PASS                  4.7 
    2 CORE                 VERIFY-DEPLOYMENT-PROVISION                                                       PASS                 2.55 
	FirstBoot : PASS 
	FirstBoot : Call Trace Verification : PASS 
	Reboot : PASS 
	Reboot : Call Trace Verification : PASS 
    3 NETWORK              NET-SET-STATIC-MAC                                                                PASS

### CODE REVIEW

**Comment by lubaihua33** [comment]
Should we need to add NetworkSecurityGroupRules into VM template?

File: XML/Other/NetworkSecurityGroupRules.xml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+<platform>
+    <linux>
+        <rule>
+            <name>CirrusLisav2SecurityRule</name>
+            <properties>
+                <description>CirrusLisav2SecurityRule</description>
+                <protocol>tcp</protocol>
+                <sourcePortRange>*</sourcePortRange>
+                <destinationPortRange>22</destinationPortRange>
+                <sourceAddressPrefix>10.0.0.4/32</sourceAddressPrefix>
```
Comment (juhlee-microsoft) [inline XML/Other/NetworkSecurityGroupRules.xml:11]: Can you post the detail information of 'sourceAddressPrefix'; what it is for, how it is working?

File: Libraries/Azure.psm1 [hunk 2]
```diff
@@ -1776,38 +1827,34 @@ Function Delete-ResourceGroup([string]$RGName, [switch]$KeepDisks, [bool]$UseExi
...
+                        RemoveAzResourcesByResourceType -ResourceType4iMatch 'Microsoft.Network/networkInterfaces'
+                        RemoveAzResourcesByResourceType -ResourceType4iMatch 'Microsoft'
                     }
-                    $CurrentResources = $unlockedResources
                     $attempts++
                 }
-                Write-LogInfo "Resources in $RGName are deleted."
-                $retValue = $true
+                if ($attempts -le 10) {
+                    Write-LogInfo "Resources in $RGName are deleted."
```
Comment (juhlee-microsoft) [inline Libraries/Azure.psm1]: The line 1852 and 1855: Can you keep the sentence in past format? We keep all logs in past format. There are a few more in other files.

**Review by juhlee-microsoft — CHANGES_REQUESTED** [review]
Logic looks good.

File: XML/Other/NetworkSecurityGroupRules.xml (hunk 1 above)
Comment (harz566) [reply XML/Other/NetworkSecurityGroupRules.xml:11]: This mask is just give an IP scope that will be allowed or denied following this rule. /32 means only the current IP itself, in real practice, the IP and mask will be replaced in the xml here as needed.

File: Libraries/Azure.psm1 (hunk 2 above)
Comment (harz566) [reply Libraries/Azure.psm1]: updated, thanks!

File: TestControllers/AzureController.psm1 [hunk 3]
```diff
@@ -48,6 +48,9 @@ Class AzureController : TestController
 		$this.TestProvider.PlatformFaultDomainCount = $this.CustomParams["PlatformFaultDomainCount"]
 		$this.TestProvider.PlatformUpdateDomainCount = $this.CustomParams["PlatformUpdateDomainCount"]
 		$this.TestProvider.EnableTelemetry = $ParamTable["EnableTelemetry"]
+		if ($this.CustomParams["EnableNSG"] -and $this.CustomParams["EnableNSG"] -eq "true") {
+			$this.TestProvider.EnableNSG = $true
+		}
```
Comment (lubaihua33) [inline TestControllers/AzureController.psm1:53]: Just use customparams can enable NSG?

Comment (harz566) [reply TestControllers/AzureController.psm1:53]: then AzureProvider and Azure.psm have logic to add nsg rules.

**Review by lubaihua33 — APPROVED** [review]

File: Libraries/Azure.psm1 [hunk 4]
```diff
@@ -797,12 +849,7 @@ Function Create-AllResourceGroupDeployments($SetupTypeData, $TestCaseData, $Dist
 		$EndPointAdded = $false
 		$role = 0
 		foreach ( $newVM in $RGXMLData.VirtualMachine) {
-			if ($newVM.RoleName) {
-				$vmName = $newVM.RoleName
-			}
-			else {
-				$vmName = Get-NewVMName -namePrefix $RGName -numberOfVMs $role
-			}
+			$vmName = $VMNames[$role]
```
Comment (LiliDeng) [inline Libraries/Azure.psm1:852]: Get-NewVMName will handle windows VM name. Windows computer name cannot be more than 15 characters long on Azure

File: Libraries/Azure.psm1 [hunk 5]
```diff
@@ -844,13 +891,7 @@ Function Create-AllResourceGroupDeployments($SetupTypeData, $TestCaseData, $Dist
...
 			$role = 0
 			foreach ( $newVM in $RGXMLData.VirtualMachine) {
-				if ($newVM.RoleName) {
-					$vmName = $newVM.RoleName
-				}
-				else {
-					$vmName = Get-NewVMName -namePrefix $RGName -numberOfVMs $role
-				}
-
+				$vmName = $VMNames[$role]
```
Comment (LiliDeng) [inline Libraries/Azure.psm1:894]: ditto

File: Libraries/Azure.psm1 [hunk 6]
```diff
@@ -1248,17 +1289,17 @@ Function Create-AllResourceGroupDeployments($SetupTypeData, $TestCaseData, $Dist
...
-						Add-Content -Value "$($indents[6]){" -Path $jsonFile
-							Add-Content -Value "$($indents[7])^publicKeys^:" -Path $jsonFile
-							Add-Content -Value "$($indents[7])[" -Path $jsonFile
-								Add-Content -Value "$($indents[8]){" -Path $jsonFile
-									Add-Content -Value "$($indents[9])^path^:^$sshPath^," -Path $jsonFile
-									Add-Content -Value "$($indents[9])^keyData^:^$sshKeyData^" -Path $jsonFile
-								Add-Content -Value "$($indents[8])}" -Path $jsonFile
-							Add-Content -Value "$($indents[7])]" -Path $jsonFile
-						Add-Content -Value "$($indents[6])}" -Path $jsonFile
+					Add-Content -Value "$($indents[6])^disablePasswordAuthentication^:true," -Path $jsonFile
```
Comment (LiliDeng) [inline Libraries/Azure.psm1:1292]: the indent is make the json file more readable.

File: Libraries/Azure.psm1 [hunk 7]
```diff
@@ -1761,7 +1802,17 @@ Function Start-DeleteResourceGroup ([string]$RGName) {
...
     return $isDeleting
 }
-Function Delete-ResourceGroup([string]$RGName, [switch]$KeepDisks, [bool]$UseExistingRG) {
+Function Delete-ResourceGroup([string]$RGName, [switch]$KeepDisks, [bool]$UseExistingRG, [array]$AllVMData) {
+    Function RemoveAzResourcesByResourceType([string] $ResourceType4iMatch) {
+        $unlockedResources | Where-Object {$_.ResourceType -imatch $ResourceType4iMatch} | ForEach-Object {
+            try {
+                $null = Remove-AzResource -ResourceGroupName $RGName -ResourceName $_.Name -ResourceType $_.ResourceType -Force -Verbose
+            } catch {
+                Write-LogErr "Failed to delete resource $($_.Name). We will try to remove it in next attempt $($attempts + 1)"
```
Comment (LiliDeng) [inline Libraries/Azure.psm1:1811]: no need pass $attempts ?

**Review by LiliDeng — CHANGES_REQUESTED** [review]

**Comment by LiliDeng** [comment]
@harz566  When I use Keep resource group, it was still deleted.

File: Libraries/Azure.psm1 (hunk 4 above)
Comment (harz566) [reply Libraries/Azure.psm1:852]: line 323

File: Libraries/Azure.psm1 (hunk 5 above)
Comment (harz566) [reply Libraries/Azure.psm1:894]: line 323

File: Libraries/Azure.psm1 (hunk 6 above)
Comment (harz566) [reply Libraries/Azure.psm1:1292]: indent will be controlled by $indents[int number], right? does it related to the indenting of the source code here itself of preparing the template file?

File: Libraries/Azure.psm1 (hunk 7 above)
Comment (harz566) [reply Libraries/Azure.psm1:1811]: it's embedded function.

**Comment by harz566** [comment]
> When I use Keep resource group, it was still delete

**Review by juhlee-microsoft — APPROVED** [review]

**Review by LiliDeng — APPROVED** [review]

**Pull request merged**